File: web/src/protocol/v0/envelope.ts
export const protocolID = "branch/0" as const;
export const maxDraftEnvelopeBytes = 64 * 1024;

const eventTypes = [
  "identity.announce",
  "rendezvous.offer",
  "rendezvous.answer",
  "route.update",
  "relay.announce",
  "capability.grant",
  "capability.revoke"
] as const;

export type EventType = (typeof eventTypes)[number];

export interface DraftEnvelope {
  readonly protocol: typeof protocolID;
  readonly id: string;
  readonly type: EventType;
  readonly sender: string;
  readonly recipient_tag: string;
  readonly created_at: number;
  readonly expires_at: number;
  readonly nonce: string;
  readonly payload: string;
  readonly signature: string;
}

export class ProtocolError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ProtocolError";
  }
}

export function isKnownEventType(value: string): value is EventType {
  return eventTypes.some((eventType) => eventType === value);
}

export function decodeDraftEnvelopeText(text: string): DraftEnvelope {
  if (new TextEncoder().encode(text).byteLength > maxDraftEnvelopeBytes) {
    throw new ProtocolError("oversized draft envelope");
  }

  const decoded: unknown = JSON.parse(text);
  return validateDraftEnvelope(decoded);
}

export function validateDraftEnvelope(value: unknown): DraftEnvelope {
  if (!isRecord(value)) {
    throw new ProtocolError("invalid draft envelope");
  }
  rejectUnknownKeys(value, [
    "protocol",
    "id",
    "type",
    "sender",
    "recipient_tag",
    "created_at",
    "expires_at",
    "nonce",
    "payload",
    "signature"
  ]);

  const envelope = {
    protocol: readLiteral(value, "protocol", protocolID),
    id: readString(value, "id"),
    type: readEventType(value, "type"),
    sender: readString(value, "sender"),
    recipient_tag: readString(value, "recipient_tag"),
    created_at: readInteger(value, "created_at"),
    expires_at: readInteger(value, "expires_at"),
    nonce: readString(value, "nonce"),
    payload: readString(value, "payload"),
    signature: readString(value, "signature")
  } satisfies DraftEnvelope;

  if (envelope.expires_at <= envelope.created_at) {
    throw new ProtocolError("expiry must be after creation");
  }

  return envelope;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function rejectUnknownKeys(record: Record<string, unknown>, keys: readonly string[]): void {
  const known = new Set(keys);
  for (const key of Object.keys(record)) {
    if (!known.has(key)) {
      throw new ProtocolError(`unknown ${key}`);
    }
  }
}

function readLiteral<T extends string>(
  record: Record<string, unknown>,
  key: string,
  expected: T
): T {
  const value = readString(record, key);
  if (value !== expected) {
    throw new ProtocolError(`unsupported ${key}`);
  }
  return expected;
}

function readString(record: Record<string, unknown>, key: string): string {
  const value = record[key];
  if (typeof value !== "string" || value.length === 0) {
    throw new ProtocolError(`missing ${key}`);
  }
  return value;
}

function readInteger(record: Record<string, unknown>, key: string): number {
  const value = record[key];
  if (typeof value !== "number" || !Number.isSafeInteger(value)) {
    throw new ProtocolError(`invalid ${key}`);
  }
  return value;
}

function readEventType(record: Record<string, unknown>, key: string): EventType {
  const value = readString(record, key);
  if (!isKnownEventType(value)) {
    throw new ProtocolError("unsupported event type");
  }
  return value;
}

File: protocol/v0/envelope.go
package v0

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
)

const (
	// ProtocolID is the accepted draft signed-event protocol identifier used by
	// the current shared fixtures.
	ProtocolID = "branch/connectivity/0"

	// MaxDraftEnvelopeBytes bounds local fixture parsing and future hostile
	// input tests. It is not a negotiated relay frame limit.
	MaxDraftEnvelopeBytes = 64 * 1024

	// MaxDraftStringBytes bounds draft string fields before the final wire
	// format defines per-field sizes.
	MaxDraftStringBytes = 1024

	// MaxDraftTimestamp is the largest integer timestamp accepted by both Go
	// and TypeScript draft conformance checks.
	MaxDraftTimestamp = 1<<53 - 1
)

// EventType names the draft signed event envelope types described in
// docs/PROTOCOL_V0.md.
type EventType string

const (
	EventIdentityAnnounce EventType = "identity.announce"
	EventRendezvousOffer  EventType = "rendezvous.offer"
	EventRendezvousAnswer EventType = "rendezvous.answer"
	EventRouteUpdate      EventType = "route.update"
	EventRelayAnnounce    EventType = "relay.announce"
	EventCapabilityGrant  EventType = "capability.grant"
	EventCapabilityRevoke EventType = "capability.revoke"
)

var (
	// ErrInvalidEnvelope reports structurally invalid draft event envelopes.
	ErrInvalidEnvelope = errors.New("invalid draft envelope")

	// ErrOversizedEnvelope reports a draft fixture that exceeds the current
	// local parsing limit.
	ErrOversizedEnvelope = errors.New("oversized draft envelope")
)

// DraftEnvelope mirrors the illustrative JSON envelope in docs/PROTOCOL_V0.md.
// The payload and signature fields are opaque strings until canonical encoding
// and cryptographic constructions are specified.
type DraftEnvelope struct {
	Protocol     string    `json:"protocol"`
	ID           string    `json:"id"`
	Type         EventType `json:"type"`
	Sender       string    `json:"sender"`
	RecipientTag string    `json:"recipient_tag"`
	CreatedAt    int64     `json:"created_at"`
	ExpiresAt    int64     `json:"expires_at"`
	Nonce        string    `json:"nonce"`
	Payload      string    `json:"payload"`
	Signature    string    `json:"signature"`
}

// DecodeDraftEnvelope decodes one draft JSON envelope with a hard byte limit.
// It intentionally performs only structural validation.
func DecodeDraftEnvelope(data []byte) (DraftEnvelope, error) {
	if len(data) > MaxDraftEnvelopeBytes {
		return DraftEnvelope{}, ErrOversizedEnvelope
	}

	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()

	var envelope DraftEnvelope
	if err := decoder.Decode(&envelope); err != nil {
		return DraftEnvelope{}, fmt.Errorf("%w: %v", ErrInvalidEnvelope, err)
	}
	if err := decoder.Decode(&struct{}{}); !errors.Is(err, io.EOF) {
		return DraftEnvelope{}, fmt.Errorf("%w: trailing data", ErrInvalidEnvelope)
	}

	if err := envelope.ValidateStructure(); err != nil {
		return DraftEnvelope{}, err
	}
	return envelope, nil
}

// ValidateStructure checks deterministic draft-envelope invariants that are
// already present in the founding protocol sketch.
func (envelope DraftEnvelope) ValidateStructure() error {
	if envelope.Protocol != ProtocolID {
		return fmt.Errorf("%w: unsupported protocol %q", ErrInvalidEnvelope, envelope.Protocol)
	}
	if !KnownEventType(envelope.Type) {
		return fmt.Errorf("%w: unsupported event type %q", ErrInvalidEnvelope, envelope.Type)
	}
	if !validDraftString(envelope.ID) {
		return fmt.Errorf("%w: missing id", ErrInvalidEnvelope)
	}
	if !validDraftString(envelope.Sender) {
		return fmt.Errorf("%w: missing sender", ErrInvalidEnvelope)
	}
	if !validDraftString(envelope.RecipientTag) {
		return fmt.Errorf("%w: missing recipient tag", ErrInvalidEnvelope)
	}
	if envelope.ExpiresAt <= envelope.CreatedAt {
		return fmt.Errorf("%w: expiry must be after creation", ErrInvalidEnvelope)
	}
	if envelope.CreatedAt < 0 || envelope.ExpiresAt > MaxDraftTimestamp {
		return fmt.Errorf("%w: timestamp out of range", ErrInvalidEnvelope)
	}
	if !validDraftString(envelope.Nonce) {
		return fmt.Errorf("%w: missing nonce", ErrInvalidEnvelope)
	}
	if !validDraftString(envelope.Payload) {
		return fmt.Errorf("%w: missing payload", ErrInvalidEnvelope)
	}
	if !validDraftString(envelope.Signature) {
		return fmt.Errorf("%w: missing signature", ErrInvalidEnvelope)
	}
	return nil
}

// KnownEventType reports whether eventType is currently listed in the draft
// protocol sketch.
func KnownEventType(eventType EventType) bool {
	switch eventType {
	case EventIdentityAnnounce,
		EventRendezvousOffer,
		EventRendezvousAnswer,
		EventRouteUpdate,
		EventRelayAnnounce,
		EventCapabilityGrant,
		EventCapabilityRevoke:
		return true
	default:
		return false
	}
}

func validDraftString(value string) bool {
	return len(value) > 0 && len(value) <= MaxDraftStringBytes
}
